### ISSUE TITLE
Incorrect type promotions with flatten

### ISSUE THREAD

**Issue by rofinn** [invenia/ParameterHandling.jl#10]
Looks like if you try to `flatten` a `deferred` function that takes an integer (probably other cases too), the returned parameters are incorrectly promoted. As a result, calling `unflatten(param)` throws an error. This seems like a more extreme case of the gotchas listed in the README, hence the issue.

Example)

```julia
julia> params, unflatten = flatten((a=rand(Float32, 2), b=deferred(fill, one(Float32), 4)))
([0.46068453788757324, 0.18974554538726807, 1.0], ParameterHandling.var"#unflatten_to_NamedTuple#10"{NamedTuple{(:a, :b),Tuple{Array{Float32,1},ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}}},ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{typeof(identity),ParameterHandling.var"#unflatten_Deferred#19"{ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}},ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{ParameterHandling.var"#unflatten_to_Real#2",ParameterHandling.var"#unflatten_Integer#1"{Int64}},Tuple{Int64,Int64},Tuple{Int64,Int64}}}},Tuple{Int64,Int64},Tuple{Int64,Int64}}}((a = Float32[0.46068454, 0.18974555], b = ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}(fill, (1.0f0, 4))), ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{typeof(identity),ParameterHandling.var"#unflatten_Deferred#19"{ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}},ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{ParameterHandling.var"#unflatten_to_Real#2",ParameterHandling.var"#unflatten_Integer#1"{Int64}},Tuple{Int64,Int64},Tuple{Int64,Int64}}}},Tuple{Int64,Int64},Tuple{Int64,Int64}}((identity, ParameterHandling.var"#unflatten_Deferred#19"{ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}},ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{ParameterHandling.var"#unflatten_to_Real#2",ParameterHandling.var"#unflatten_Integer#1"{Int64}},Tuple{Int64,Int64},Tuple{Int64,Int64}}}(ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}(fill, (1.0f0, 4)), ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{ParameterHandling.var"#unflatten_to_Real#2",ParameterHandling.var"#unflatten_Integer#1"{Int64}},Tuple{Int64,Int64},Tuple{Int64,Int64}}((ParameterHandling.var"#unflatten_to_Real#2"(), ParameterHandling.var"#unflatten_Integer#1"{Int64}(4)), (1, 0), (1, 1)))), (2, 1), (2, 3))))

julia> typeof(params)
Array{Float64,1}

julia> unflatten(params)
ERROR: MethodError: Cannot `convert` an object of type
  ParameterHandling.Deferred{typeof(fill){},Tuple{Float64,Int64{}}} to an object of type
  ParameterHandling.Deferred{typeof(fill){},Tuple{Float32,Int64{}}}
Closest candidates are:
  convert(::Type{T}, ::T) where T at essentials.jl:171
  ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}(::Any, ::Any) where {Tf, Targs} at /Users/rory/.julia/packages/ParameterHandling/W3SDV/src/parameters.jl:140
Stacktrace:
 [1] convert(::Type{Tuple{ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}}}, ::Tuple{ParameterHandling.Deferred{typeof(fill),Tuple{Float64,Int64}}}) at ./essentials.jl:310 (repeats 2 times)
 [2] Tuple{Array{Float32,1},ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}}(::Tuple{Array{Float64,1},ParameterHandling.Deferred{typeof(fill),Tuple{Float64,Int64}}}) at ./tuple.jl:225
 [3] NamedTuple{(:a, :b),Tuple{Array{Float32,1},ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}}}(::Tuple{Array{Float64,1},ParameterHandling.Deferred{typeof(fill),Tuple{Float64,Int64}}}) at ./namedtuple.jl:90
 [4] (::ParameterHandling.var"#unflatten_to_NamedTuple#10"{NamedTuple{(:a, :b),Tuple{Array{Float32,1},ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}}}},ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{typeof(identity),ParameterHandling.var"#unflatten_Deferred#19"{ParameterHandling.Deferred{typeof(fill),Tuple{Float32,Int64}},ParameterHandling.var"#unflatten_to_Tuple#8"{Tuple{ParameterHandling.var"#unflatten_to_Real#2",ParameterHandling.var"#unflatten_Integer#1"{Int64}},Tuple{Int64,Int64},Tuple{Int64,Int64}}}},Tuple{Int64,Int64},Tuple{Int64,Int64}}})(::Array{Float64,1}) at /Users/rory/.julia/packages/ParameterHandling/W3SDV/src/flatten.jl:73
 [5] top-level scope at REPL[41]:1
```

**Comment by willtebbutt** [comment]
Is this resolved by #13 , or is there more to do?

**Comment by rofinn** [comment]
Yeah, it should be.
